File: blackjack.py
import random
N=int(input('Enter number of decks: '))
#array holding the total cards in the game 
totalCards=[]
#deck representation
deck=['2♥','3♥','4♥','5♥','6♥','7♥','8♥','9♥','10♥','J♥','Q♥','K♥','A♥'
     ,'2♦','3♦','4♦','5♦','6♦','7♦','8♦','9♦','10♦','J♦','Q♦','K♦','A♦'
     ,'2♣','3♣','4♣','5♣','6♣','7♣','8♣','9♣','10♣','J♣','Q♣','K♣','A♣'
     ,'2♠','3♠','4♠','5♠','6♠','7♠','8♠','9♠','10♠','J♠','Q♠','K♠','A♠'
]
#adding number of decks in total cards
for b in range(N):
    for i in range(len(deck)):
        totalCards.append(deck[i])

# print(len(totalCards))
# print(totalCards)

#shuffling total cards
random.shuffle(totalCards)

# print(totalCards)
playerMoney=100
flag=True
while(flag):
    if (len(totalCards)>int(0.25*N*52)):
        print('Player Money: ',playerMoney,' - Cards Left: ',len(totalCards))
        print('Select your bet in [1,{}]:'.format(playerMoney))
        bet=input()
        print('Dealer cards: ',totalCards[0],'??')
        dealerCards=[totalCards[0],totalCards[1]]
        print('Player Cards: ',totalCards[2],totalCards[3])
        game=True
        playerScore=0
        dealerScore=0
        if totalCards[2][0]=='A' or totalCards[2][0]=='J' or totalCards[2][0]=='K' or totalCards[2][0]=='Q':
            if totalCards[2][0]=='A':
                playerScore+=11
            else:
                playerScore+=10
        else:
            playerScore+=int(totalCards[2][0])
        if totalCards[3][0]=='A' or totalCards[3][0]=='J' or totalCards[3][0]=='K' or totalCards[3][0]=='Q':
            if totalCards[3][0]=='A':
                playerScore+=11
            else:
                playerScore+=10
        else:
            playerScore+=int(totalCards[3][0])

        if totalCards[0][0]=='A' or totalCards[0][0]=='J' or totalCards[0][0]=='K' or totalCards[0][0]=='Q':
            if totalCards[0][0]=='A':
                dealerScore+=11
            else:
                dealerScore+=10
        else:
            dealerScore+=int(totalCards[0][0])
        if totalCards[1][0]=='A' or totalCards[1][0]=='J' or totalCards[1][0]=='K' or totalCards[1][0]=='Q':
            if totalCards[1][0]=='A':
                dealerScore+=11
            else:
                dealerScore+=10
        else:
            dealerScore+=int(totalCards[1][0])
        totalCards=totalCards[4:]
        if playerScore==21 and dealerScore!=21:
            print('You win bet...')
            playerMoney+=int(bet)
            game=False
            break
        if dealerScore==21 and playerScore!=21:
            print('You lose bet...')
            playerMoney-=int(bet)
            game=False
            break
        if dealerScore==21 and playerScore==21:
            print('No body wins - Tie')
            game=False
            break
        if game:
            hit=input('Do you want ot hit? [y=yess/n=no] ')
            if hit=='y':
                bust=False
                while not bust:
                    if totalCards[0][0]=='A' or totalCards[0][0]=='K' or totalCards[0][0]=='J' or totalCards[0][0]=='Q':
                        if totalCards[0][0]=='A':
                            if playerScore+11<21:
                                playerScore+=11
                            else:
                                playerScore+=1
                        else:
                            playerScore+=10
                    else:
                        playerScore+=int(totalCards[0][0])
                    print('You drew card: ',totalCards[0],' your sum is: ',playerScore)
                    totalCards=totalCards[1:]
                    if playerScore<21:
                        hit=input('Do you want ot hit? [y=yess/n=no] ')
                        if hit=='y':
                            bust=False
                        if hit=='n':
                            bust=True
                    if playerScore>21:
                        bust=True
    if (len(totalCards)>int(0.25*N*52)):
        print('Dealer cards: ',dealerCards[0],dealerCards[1])
        bust=False
        while not bust:
            if totalCards[0][0]=='A' or totalCards[0][0]=='K' or totalCards[0][0]=='J' or totalCards[0][0]=='Q':
                if totalCards[0][0]=='A':
                    if dealerCards+11<17:
                        dealerScore+=11
                    else:
                        dealerScore+=1
                else:
                    dealerScore+=10
            else:
                dealerScore+=int(totalCards[0][0])
            print('Dealer drew card: ',totalCards[0],' the sum is: ',dealerScore)
            if dealerScore>playerScore:
                print('You lose bet...')
                playerMoney-=int(bet)
                bust=True
                break
            elif playerScore>dealerScore:
                print('You win bet...')
                playerMoney+=int(bet)
                bust=True
                break
            elif playerScore==dealerScore:
                print('No body wins - Tie')
                bust=True
                break
            totalCards=totalCards[1:]
            if dealerScore>=17:
                bust=True

    else:
        flag=False

    

    x=input('\nDo you want to continue? [y=yess/n=no] ')
    if x=='y':
        flag=True
    if x=='n':
        flag=False
print('Player Money: ',playerMoney,' - Cards Left: ',len(totalCards))
print('Goodbye')

File: BlackJack.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
int money=100;
int noofdecks;
int totalcards;
int shuff(int cards[])
{
	
	int t;
	int i;
	int desk[52];	
	for (i=0;i<52;i++)
		desk[i] = (i/13+3)*100 + i%13 + 1;

	srand(time(NULL));
	for (i = 0; i < 52; i++)
	{
		do
		{
			t = rand() % 52;
		} while (desk[t] == 0);
		cards[i] = desk[t];
		desk[t] = 0;
	}
	
	return 0;
}

int convert_jkq(int a)
{
	if ((a%100==11) ||(a%100==12) ||(a%100==13)) return (a/100)*100+10;
	else return a;
}

void pic(int num)
{
	char fl;
	int po_num;
	
	fl = num / 100;
	po_num = num % 100;
	switch (po_num)
	{
		case 1: 
		{
			printf("A%c",fl);
			break;
		}
		case 2:
		case 3:
		case 4:
		case 5:
		case 6:
		case 7:
		case 8:
		case 9:
		case 10:
		{
			
			printf("%2d%c",po_num,fl);
			break;
		}
		case 11:
		{
			printf("J%c",fl);
			break;
		}
		case 12:
		{
			printf("Q%c",fl);
			break;
		}
		case 13:
		{
			printf("K%c",fl);
			break;
		}

	}
}

int play(void)
{
	int i;
	int psum=0;
	int bsum=0;
	int pcards[5]={0};
	int bcards[5]={0};
	int cards[52];
	char go_on;
	char d;
	shuff(cards);

	//give the cards
	pcards[0]=cards[0];
	pcards[1]=cards[1];
	bcards[0]=cards[2];
	bcards[1]=cards[3];
	
	//the 2 cards player get
	printf("Player Money: %d - Cards Left: %d",money,totalcards);
	printf("\nSelect your bet in [1,%d]: ",money);
	int bet=0;
	scanf("%d",&bet);
	printf("Dealer cards: ");
	pic(bcards[0]);
	printf("??\nPlayer cards: ");
	pic(pcards[0]);
	pic(pcards[1]);
	totalcards=totalcards-4;
	i=0;
	for (i=0; i<2; i++)
	{
		if (pcards[i]%100 == 1)
		{			
			if (psum+11<21)
			{
				psum = psum + 11;
			}
			else
			{
				psum = psum +1;
			}
		}
		else if (convert_jkq(pcards[i]) %100 ==10) psum = psum + 10;
		else psum = psum + pcards[i]%100;
		
		if (psum > 21)
		{
			printf("\nYou lose bet...\n");
			money=money-bet;
			return 1;
		}
		else if (psum == 21)
		{
			printf("\nYou win bet...\n");
			money=money+bet;
			return 0;
		}
	}
	
	//whether player get another cards
	i=0;
	for (i=0; i<3; i++)
	{
		char j = 'n';
		
		printf("\nDo you want to hit? [y=yess/n=no] ");
		do{
			j = getchar();
		} while (j!='y' &&j!='n');
		
		if (j=='y')
		{
			pcards[i+2]=cards[i+4];
			printf("You drew card: ");
			pic(pcards[i+2]);
			totalcards=totalcards-1;
			if (pcards[i+2]%100 == 1)
			{
				
				if (psum+11 <21)
				{
					psum = psum + 11;
				}
				else
				{
					psum = psum +1;
				}
			}
			else if (convert_jkq(pcards[i+2]) %100 ==10) psum = psum + 10;
			else psum = psum + pcards[i+2]%100;
			
			if (psum > 21)
			{
				printf("  your sum is: %d\n\n",psum);
				printf("You lose bet...\n");
				money=money-bet;
				return 1;
			}
			else if (psum == 21)
			{
				printf(" your sum is: %d\n\n",psum);
				printf("You win bet...\n");
				money=money+bet;
				return 0;
			}		
			else printf(" your sum is: %d\n\n",psum);
		}
		else 
		{
			break;
		}
	}
	if (i == 3)
	{
		printf("You win bet...\n");
		money=money+bet;
		return 0;
	}
	
	printf("Dealer cards: ");
	pic(bcards[0]);
	pic(bcards[1]);

	if (bcards[0]%100 + bcards[1]%100 == 2)
	{
		bsum=12; //two A cards
		printf(" the sum is: %d\n\n", bsum);
	}
	else if ((convert_jkq(bcards[0]))%100 + (convert_jkq(bcards[1]))%100 ==1)
	{
		bsum=21;
		printf("You lose bet...\n");
		money=money-bet;
		return 1;
	}
	else if (bcards[0]%100==1 || bcards[1]%100==1)
	{
		bsum=(bcards[0]+bcards[1])%100+(rand()%2)*10;
		printf(" the sum is: %d\n\n", bsum);
	}
	else
	{
		bsum = (convert_jkq(bcards[0]))%100 + (convert_jkq(bcards[1]))%100;
		printf(" the sum is: %d\n\n", bsum);
	}
	
	//whether Dealer get another cards until bsum>16
	for (i=0; i<3 && bsum<17; i++)
	{
		bcards[i+2]=cards[i+7];
		printf("Dealer drew card: ");
		pic(bcards[i+2]);
		totalcards=totalcards-1;
		if (bcards[i+2]%100 == 1)
		{
			if (bsum+11 <= 21)
			{
				bsum = bsum+11;
				printf(" the sum is:%d\n\n", bsum);
			}
			else
			{
				bsum = bsum+1;
				printf(" the sum is:%d\n\n", bsum);
			}
		}
		else
		{
			bsum = bsum + convert_jkq(bcards[i+2])%100;
			printf(" the sum is:%d\n\n", bsum);
		}
	}
	if (i == 3)
	{
		printf("You lose bet...\n");
		money=money-bet;
		return 1;
	}
	
	//the last step
	if (bsum>21 || psum>bsum)
	{
		printf("You win bet...\n");
		money=money+bet;
		return 0;
	}
	else if (psum == bsum)
	{
		printf("No body wins - Tie\n");
		return 3;
	}
	else if (psum < bsum)
	{
		printf("You lose bet...\n");
		money=money-bet;
		return 1;
	}
		
	return 3;
}

int main(void)
{
	char again;
	printf("Enter number of decks: ");
	scanf("%d",&noofdecks);
	totalcards=52*noofdecks;
	play();
	printf("\nDo you want to contiue? [y=yes/n=no] ");
	do{
		again = getchar();
	} while (again!='y' && again!='n');
	while (again=='y'){
		play();
		printf("\nDo you want to contiue? [y=yes/n=no] ");
			do{
				again = getchar();
			} while (again!='y' && again!='n');
	};
	printf("Player Money: %d - Cards Left: %d\nGoodbye",money,totalcards);
	return 0;
}
